cardano-api/src/Cardano/Api/ProtocolParametersUpdate.hs

```diff
@@ -0,0 +1,674 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | The various Cardano protocol parameters, including:
+--
+-- * the current values of updateable protocol parameters: 'ProtocolParameters'
+-- * updates to protocol parameters: 'ProtocolParametersUpdate'
+-- * update proposals that can be embedded in transactions: 'UpdateProposal'
+-- * parameters fixed in the genesis file: 'GenesisParameters'
+--
+module Cardano.Api.ProtocolParametersUpdate (
+    -- * Updates to the protocol paramaters
+    ProtocolParametersUpdate(..),
+
+    -- * Update proposals to change the protocol paramaters
+    UpdateProposal(..),
+    makeShelleyUpdateProposal,
+
+    -- * Internal conversion functions
+    toShelleyPParamsUpdate,
+    toShelleyProposedPPUpdates,
+    toShelleyUpdate,
+    toUpdate,
+    fromShelleyPParamsUpdate,
+    fromShelleyProposedPPUpdates,
+    fromShelleyUpdate,
+
+    -- * Data family instances
+    AsType(..)
+  ) where
+
+import           Prelude
+
+import           Data.Aeson (FromJSON (..), ToJSON (..), object, withObject, (.:), (.=))
+import qualified Data.Aeson.Types as Aeson
+import qualified Data.ByteString.Lazy.Char8 as LBS
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (mapMaybe)
+import           Data.Text (Text)
+import           Data.Word (Word64)
+import           Numeric.Natural
+
+import           Control.Monad
+
+import qualified Cardano.Binary as CBOR
+
+import qualified Cardano.Ledger.Era as Ledger
+import qualified Cardano.Ledger.Shelley.Constraints as Shelley
+import           Ouroboros.Consensus.Shelley.Eras (StandardShelley)
+import           Ouroboros.Consensus.Shelley.Protocol.Crypto (StandardCrypto)
+
+import           Cardano.Api.Address
+import           Cardano.Api.Eras
+import           Cardano.Api.HasTypeProxy
+import           Cardano.Api.Hash
+import           Cardano.Api.KeysByron
+import           Cardano.Api.KeysShelley
+import           Cardano.Api.ProtocolParameters
+import           Cardano.Api.Script
+import           Cardano.Api.SerialiseCBOR
+import           Cardano.Api.SerialiseTextEnvelope
+import           Cardano.Api.Value
+
+import qualified Cardano.Ledger.Alonzo.Language as Alonzo
+import qualified Cardano.Ledger.Alonzo.Scripts as Alonzo
+import qualified PlutusCore.Evaluation.Machine.ExBudgeting as Plutus
+import           Shelley.Spec.Ledger.BaseTypes (maybeToStrictMaybe, strictMaybeToMaybe)
+import qualified Shelley.Spec.Ledger.BaseTypes as Shelley
+import qualified Shelley.Spec.Ledger.PParams as Shelley
+
+-- ----------------------------------------------------------------------------
+-- Updates to the protocol paramaters
+--
+
+-- | The representation of a change in the 'ProtocolParameters'.
+--
+data ProtocolParametersUpdate =
```

Comment:
Note how this type does _not_ have the era param, and yet it contains essentially identical info to the `ProtocolParameters` type. That's because neither of them need the era type param.

BTW, if we avoid the truly massive constructor functions in the main `ProtocolParameters` module as I suggest above, then we might not need to split this module out and that might keep things simpler.